xarray/core/indexing.py

```diff
@@ -1291,9 +1302,27 @@ def __getitem__(self, indexer):
 
         return result
 
-    def transpose(self, order):
+    def transpose(self, order) -> pd.Index:
         return self.array  # self.array should be always one-dimensional
 
-    def __repr__(self):
+    def __repr__(self) -> str:
         return ('%s(array=%r, dtype=%r)'
                 % (type(self).__name__, self.array, self.dtype))
+
+    def copy(self, deep: bool = True) -> 'PandasIndexAdapter':
+        # Not the same as just writing `self.array.copy(deep=deep)`, as
+        # shallow copies of the underlying numpy.ndarrays become deep ones
+        # upon pickling
```

Comment:
I don't quite follow this comment -- how does pickling relate to this new method?

Reply:
``pandas.Index.copy(deep=False)`` creates new identical views of the underlying numpy arrays.
A numpy view becomes a real, deep-copied numpy array upon pickling. This is by design to prevent people from accidentally sending a huge numpy array over network or IPC when they just want to send a slice of it. Crucially though, this (IMHO improvable) design makes no distinction when the base array is the same size or smaller than the view.

Back to the method at hand: if two xarray.DataArray/Dataset share the same underlying pandas.Index, which is the whole intent of copy(deep=False), when they are pickled together they continue sharing the same data. This is already true when you shallow-copy xarray.Variable.
But if instead of
```
array = self.array.copy(deep=True) if deep else self.array
```
we wrote
```
array = self.array.copy(deep=deep)
```
which would be tempting as it is much more readable, everything would be the same initially, _until you pickle the original index and its shallow copy together_, at which point the copy would automatically and silently become a deep one, causing double RAM/disk usage upon storing/unpickling.

```
>>> idx = xarray.IndexVariable('x', list(range(500000)))._data.array
>>> len(pickle.dumps((idx, idx)))                                                                                                       
4000281
>>> len(pickle.dumps((idx, idx.copy(deep=False))))                                                                                     
8000341
```

This was a real problem I faced a couple of years ago where I had to dump to disk (for forensic purposes)  about 10,000 intermediate steps of a Monte Carlo simulation, each step being a DataArray or Dataset. The data variables were all dask-backed, so they were fine. But among the indices I had a 'scenario' dimension with 500,000 points, dtype='<U13'.

Before pickling: 13 * 500,000 = 6.2 MB; the 50,000 shallow copies of it being just views
After pickling: 13 * 500,000 * 10,000 = I needed a new hard disk!

This issue was from a couple of years ago and has since been fixed. My (apparently overcomplicated) code above prevents it from showing up again.

P.S. this, by the way, is a big problem with ``xarray.align`` when you are dealing with mixed numpy+dask arrays

```
>>> a = xarray.DataArray(da.ones(500000, chunks=10000), dims=['x'])                                                                     
>>> b = xarray.concat([a] + [xarray.DataArray(i) for i in range(1000)], dim='y')                                                        
>>> c = b.sum()
>>> c.compute()
<xarray.DataArray ()>
array(2.497505e+11)
# Everything is fine so far (not very fast due to the dask chunks of size 1 but still)...
>>> client = distributed.Client()
>>> c.compute()
```
Watch the RAM usage of your dask client rocket to 8 GB, and several GBs worker-side too.
This will take a while because, client side, you just created 3.8 GB (500000 * 8 * 1000) worth of pickle file and are now sending it over the network.

Reply:
Thanks for the explanation, this makes complete sense.